From 03a2e83e35b786cce2354e643648e00543bf3082 Mon Sep 17 00:00:00 2001
From: Devi Priya <devipriy@qti.qualcomm.com>
Date: Thu, 27 Aug 2026 15:13:21 +0530
Subject: [PATCH] drivers: thermal: tsens: Expose maximum TSENS temperature as
 thermal zone

The TSENS hardware supports reporting the maximum temperature across
all sensors. Add a dedicated thermal zone to expose this aggregated
temperature value, allowing userspace to monitor the highest
TSENS temperature at any point in time.

Change-Id: I325550a0e0708ad72d57ae67e4b9bdf2ac95ec93
Signed-off-by: Devi Priya <devipriy@qti.qualcomm.com>
(cherry picked from commit a1e895e895907659b923ec2fd3838e3a701b5590)
---
 drivers/thermal/qcom/tsens-v2.c | 20 ++++++-
 drivers/thermal/qcom/tsens.c    | 96 +++++++++++++++++++++++++++++++++
 drivers/thermal/qcom/tsens.h    | 20 +++++++
 3 files changed, 135 insertions(+), 1 deletion(-)

diff --git a/drivers/thermal/qcom/tsens-v2.c b/drivers/thermal/qcom/tsens-v2.c
index 7a1f6f0e724b..1e655559af08 100644
--- a/drivers/thermal/qcom/tsens-v2.c
+++ b/drivers/thermal/qcom/tsens-v2.c
@@ -68,11 +68,28 @@ static struct tsens_features ipq5332_feat = {
 	.combo_int	= 1,
 	.adc		= 0,
 	.srot_split	= 1,
+	.has_max_temp_reg = 1,
+	.has_max_temp_valid_bit = 0,
+	.max_temp_reg_off = 0x00e0,
 	.max_sensors	= 16,
 	.trip_min_temp	= 0,
 	.trip_max_temp	= 204000,
 };
 
+static struct tsens_features ipq9650_feat = {
+	.ver_major		= VER_2_X_NO_RPM,
+	.crit_int		= 1,
+	.combo_int		= 1,
+	.adc			= 0,
+	.srot_split		= 1,
+	.has_max_temp_reg	= 1,
+	.has_max_temp_valid_bit	= 1,
+	.max_temp_reg_off	= 0x00e8,
+	.max_sensors		= 16,
+	.trip_min_temp		= 0,
+	.trip_max_temp		= 204000,
+};
+
 static const struct reg_field tsens_v2_regfields[MAX_REGFIELDS] = {
 	/* ----- SROT ------ */
 	/* VERSION */
@@ -278,6 +295,7 @@ struct tsens_plat_data data_ipq8074 = {
 static const struct tsens_ops ops_ipq5332 = {
 	.init		= init_tsens_v2_no_rpm,
 	.get_temp	= get_temp_tsens_valid,
+	.get_max_temp	= get_max_temp_tsens_valid,
 	.calibrate	= tsens_v2_calibration,
 };
 
@@ -309,7 +327,7 @@ struct tsens_plat_data data_ipq9650 = {
 	.num_sensors	= 11,
 	.ops		= &ops_ipq5332,
 	.hw_ids		= (unsigned int []){5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15},
-	.feat		= &ipq5332_feat,
+	.feat		= &ipq9650_feat,
 	.fields		= tsens_v2_regfields,
 };
 
diff --git a/drivers/thermal/qcom/tsens.c b/drivers/thermal/qcom/tsens.c
index 157f471ccf38..ed0c3f21c863 100644
--- a/drivers/thermal/qcom/tsens.c
+++ b/drivers/thermal/qcom/tsens.c
@@ -984,6 +984,32 @@ int get_temp_tsens_valid(const struct tsens_sensor *s, int *temp)
 	return 0;
 }
 
+int get_max_temp_tsens_valid(const struct tsens_sensor *s, int *temp)
+{
+	struct tsens_priv *priv = s->priv;
+	u32 valid, sensor_id;
+	int ret;
+
+	if (priv->feat->has_max_temp_valid_bit) {
+		/*
+		 * Valid bit is 0 for 6 AHB clock cycles, so poll until the
+		 * valid bit gets set before reading the temperature
+		 */
+		ret = regmap_field_read_poll_timeout(priv->rf[TSENS_MAX_TEMP_VALID], valid,
+						     valid, 1, 20 * USEC_PER_MSEC);
+		if (ret)
+			return ret;
+	}
+
+	*temp = tsens_hw_to_mC(s, TSENS_MAX_TEMP);
+
+	ret = regmap_field_read(priv->rf[TSENS_MAX_TEMP_SENSOR_ID], &sensor_id);
+	if (!ret)
+		dev_dbg(priv->dev, "max temp reported by sensor hw_id %u\n", sensor_id);
+
+	return 0;
+}
+
 int get_temp_common(const struct tsens_sensor *s, int *temp)
 {
 	struct tsens_priv *priv = s->priv;
@@ -1270,6 +1296,39 @@ int __init init_common(struct tsens_priv *priv)
 		regmap_field_write(priv->rf[CC_MON_MASK], 1);
 	}
 
+	if (priv->feat->has_max_temp_reg) {
+		struct reg_field rf_max_temp =
+			REG_FIELD(priv->feat->max_temp_reg_off, 0, 11);
+		struct reg_field rf_sensor_id =
+			REG_FIELD(priv->feat->max_temp_reg_off, 12, 15);
+
+		priv->rf[TSENS_MAX_TEMP] = devm_regmap_field_alloc(dev, priv->tm_map,
+								   rf_max_temp);
+		if (IS_ERR(priv->rf[TSENS_MAX_TEMP])) {
+			ret = PTR_ERR(priv->rf[TSENS_MAX_TEMP]);
+			goto err_put_device;
+		}
+
+		priv->rf[TSENS_MAX_TEMP_SENSOR_ID] =
+			devm_regmap_field_alloc(dev, priv->tm_map, rf_sensor_id);
+		if (IS_ERR(priv->rf[TSENS_MAX_TEMP_SENSOR_ID])) {
+			ret = PTR_ERR(priv->rf[TSENS_MAX_TEMP_SENSOR_ID]);
+			goto err_put_device;
+		}
+
+		if (priv->feat->has_max_temp_valid_bit) {
+			struct reg_field rf_valid =
+				REG_FIELD(priv->feat->max_temp_reg_off, 16, 16);
+
+			priv->rf[TSENS_MAX_TEMP_VALID] =
+				devm_regmap_field_alloc(dev, priv->tm_map, rf_valid);
+			if (IS_ERR(priv->rf[TSENS_MAX_TEMP_VALID])) {
+				ret = PTR_ERR(priv->rf[TSENS_MAX_TEMP_VALID]);
+				goto err_put_device;
+			}
+		}
+	}
+
 	spin_lock_init(&priv->ul_lock);
 
 	/* VER_0 interrupt doesn't need to be enabled */
@@ -1289,6 +1348,17 @@ static int tsens_get_temp(struct thermal_zone_device *tz, int *temp)
 	return priv->ops->get_temp(s, temp);
 }
 
+static int tsens_get_max_temp(struct thermal_zone_device *tz, int *temp)
+{
+	struct tsens_sensor *s = thermal_zone_device_priv(tz);
+	struct tsens_priv *priv = s->priv;
+
+	if (!priv->ops->get_max_temp)
+		return -EOPNOTSUPP;
+
+	return priv->ops->get_max_temp(s, temp);
+}
+
 static int  __maybe_unused tsens_suspend(struct device *dev)
 {
 	struct tsens_priv *priv = dev_get_drvdata(dev);
@@ -1388,6 +1458,14 @@ static const struct thermal_zone_device_ops tsens_of_ops = {
 #endif
 };
 
+/*
+ * The virtual "max temperature" sensor only exposes a HW-latched value
+ * and has no thresholds or trips of its own to program.
+ */
+static const struct thermal_zone_device_ops tsens_max_of_ops = {
+	.get_temp = tsens_get_max_temp,
+};
+
 static int tsens_register_irq(struct tsens_priv *priv, char *irqname,
 			      irq_handler_t thread_fn)
 {
@@ -1448,6 +1526,24 @@ static int tsens_register(struct tsens_priv *priv)
 		devm_thermal_add_hwmon_sysfs(priv->dev, tzd);
 	}
 
+	if (priv->feat->has_max_temp_reg && priv->ops->get_max_temp) {
+		priv->max_sensor = devm_kzalloc(priv->dev, sizeof(*priv->max_sensor),
+						GFP_KERNEL);
+		if (!priv->max_sensor)
+			return -ENOMEM;
+
+		priv->max_sensor->priv = priv;
+		priv->max_sensor->hw_id = MAX_SENSOR_HW_ID;
+
+		tzd = devm_thermal_of_zone_register(priv->dev, MAX_SENSOR_HW_ID,
+						    priv->max_sensor,
+						    &tsens_max_of_ops);
+		if (!IS_ERR(tzd)) {
+			priv->max_sensor->tzd = tzd;
+			devm_thermal_add_hwmon_sysfs(priv->dev, tzd);
+		}
+	}
+
 	/* VER_0 require to set MIN and MAX THRESH
 	 * These 2 regs are set using the:
 	 * - CRIT_THRESH_0 for MAX THRESH hardcoded to 120°C
diff --git a/drivers/thermal/qcom/tsens.h b/drivers/thermal/qcom/tsens.h
index aab61ee3d32c..6bd76b077771 100644
--- a/drivers/thermal/qcom/tsens.h
+++ b/drivers/thermal/qcom/tsens.h
@@ -21,6 +21,7 @@
 #define THRESHOLD_MIN_ADC_CODE	0x0
 
 #define MAX_SENSORS 16
+#define MAX_SENSOR_HW_ID	129
 
 #include <linux/interrupt.h>
 #include <linux/thermal.h>
@@ -81,6 +82,7 @@ struct tsens_ops {
 	int (*calibrate)(struct tsens_priv *priv);
 	int (*get_temp)(const struct tsens_sensor *s, int *temp);
 	/* optional callbacks */
+	int (*get_max_temp)(const struct tsens_sensor *s, int *temp);
 	int (*enable)(struct tsens_priv *priv, int i);
 	void (*disable)(struct tsens_priv *priv);
 	int (*suspend)(struct tsens_priv *priv);
@@ -463,6 +465,11 @@ enum regfield_ids {
 	CC_MON_CLEAR,
 	CC_MON_MASK,
 
+	/* HW-computed maximum temperature across all sensors */
+	TSENS_MAX_TEMP,
+	TSENS_MAX_TEMP_SENSOR_ID,
+	TSENS_MAX_TEMP_VALID,
+
 	MIN_STATUS_0,		/* MIN threshold violated */
 	MIN_STATUS_1,
 	MIN_STATUS_2,
@@ -509,9 +516,15 @@ enum regfield_ids {
  * @srot_split: does the IP neatly splits the register space into SROT and TM,
  *              with SROT only being available to secure boot firmware?
  * @has_watchdog: does this IP support watchdog functionality?
+ * @has_max_temp_reg: does the IP have a HW register that latches the max
+ *                     temperature across all sensors?
+ * @has_max_temp_valid_bit: does the MAX_TEMP register have a valid
+ *                     status bit that must be polled before reading?
  * @max_sensors: maximum sensors supported by this version of the IP
  * @trip_min_temp: minimum trip temperature supported by this version of the IP
  * @trip_max_temp: maximum trip temperature supported by this version of the IP
+ * @max_temp_reg_off: TM offset of the MAX_TEMP register;
+ *                    must be set whenever has_max_temp_reg is set
  */
 struct tsens_features {
 	unsigned int ver_major;
@@ -520,9 +533,12 @@ struct tsens_features {
 	unsigned int adc:1;
 	unsigned int srot_split:1;
 	unsigned int has_watchdog:1;
+	unsigned int has_max_temp_reg:1;
+	unsigned int has_max_temp_valid_bit:1;
 	unsigned int max_sensors;
 	int trip_min_temp;
 	int trip_max_temp;
+	u32 max_temp_reg_off;
 };
 
 /**
@@ -569,6 +585,7 @@ struct tsens_context {
  * @debug_root: pointer to debugfs dentry for all tsens
  * @debug: pointer to debugfs dentry for tsens controller
  * @sensor: list of sensors attached to this device
+ * @max_sensor: virtual sensor exposing the HW-computed max temperature
  */
 struct tsens_priv {
 	struct device			*dev;
@@ -589,6 +606,8 @@ struct tsens_priv {
 	struct dentry			*debug_root;
 	struct dentry			*debug;
 
+	struct tsens_sensor		*max_sensor;
+
 	struct tsens_sensor		sensor[];
 };
 
@@ -637,6 +656,7 @@ int tsens_calibrate_common(struct tsens_priv *priv);
 void compute_intercept_slope(struct tsens_priv *priv, u32 *pt1, u32 *pt2, u32 mode);
 int init_common(struct tsens_priv *priv);
 int get_temp_tsens_valid(const struct tsens_sensor *s, int *temp);
+int get_max_temp_tsens_valid(const struct tsens_sensor *s, int *temp);
 int get_temp_common(const struct tsens_sensor *s, int *temp);
 
 /* TSENS target */
-- 
2.34.1

